### ISSUE TITLE
Use object to describe merge descisions

### ISSUE THREAD

**Issue by vidartf** [jupyter/nbdime#74]
Motivated by a need to communicate the merge results/conflicts to the mergetool, I would propose to have the merge algorithm build a structure which defines its decisions. A gist outlining a proposal for such a format can be found here: https://gist.github.com/vidartf/5ff461d224583772ff799732956d8f07

This decision structure could then be applied to the base directly to produce the merged file or sent to the merge tool.

Advantages include:
 - Mergetool gets all the information it needs:
    - Which diffs are in conflict with each others.
    - Can easily show local/remote diffs aligned with merged output.
    - Can show the source of all difference in merged vs. base (i.e. color highlighting to say "this change comes from local, and this change comes from remote", which is good for context).
    - Has the original diffs easily accessible if the user wants to overrule/reset default decisions.
 - In the merge routine, the decision making is separated from how those decisions are applied to the notebook (e.g. inserting conflict markers). This should also help compartmentalize the code for easier maintenance and unit testing.

Input would be very welcome on:
 - Are there any cases where this will break down / cause inefficiencies?
 - Should there be any changes to the proposed format for describing the decisions?

**Comment by martinal** [comment]
You convinced me of the general concept yesterday and I'm still with you.

Some input:

For the simple case, action could also be "either" when local and remote diffs are equal. Not important for the patching, but useful for the display?

I believe your actions list for the resolved case covers all relevant cases from the strategies list in nbdime/merging/notebooks.py.

I'm not sure if we need both type and action, maybe they can be combined:
- type=simple corresponds to action="local" or "remote" or "either"
- type=conflict corresponds to action=None (or maybe use "conflict" or "undecided")

**Comment by martinal** [comment]
Also as we talked about yesterday, which may not be clear for other readers here, the merge algorithm produces this merge data structure but does not apply it in case of conflict (i.e. if any item here has action=None) but rather sends the base document and this structure for the GUI to apply. This seems to be a clean separation of concerns, and we avoid doing some patching here, some patching there.

**Comment by martinal** [comment]
Basically when this structure is passed around, it represents different things:

- from the initial merge it represents the safe easy to resolve actions
- from the autoresolve stage it represents the opinionated resolutions
- when the gui gets this it can treat it as the recommendations to present to the user
- as the user works with the gui, this structure represents the resolutions the user chooses

**Comment by minrk** [comment]
Sounds like a solid proposal, I don't see any immediate issues. I agree that we could probably merge 'type' into 'action', unless I'm misunderstanding something.

**Comment by martinal** [comment]
We just agreed to drop type but add a boolean field "conflict", which if true means that "action" is a suggestion that needs manual inspection.

**Comment by vidartf** [comment]
gist updated (https://gist.github.com/vidartf/5ff461d224583772ff799732956d8f07)

**Comment by vidartf** [comment]
Added a PR for a JSON schema describing this format in #75.

**Comment by vidartf** [comment]
A possible addition to this would be to add an action "recurse", which will then include a sublist of merge decisions that replaces the higher level merge decision by recursion. An example:

Two list elements have both been inserted simultaneously at the same key, one from local, and one from remote. If these two elements are found to be very similar by inspection of the edit distance, the auto resolve algorithm can choose to recurse into these, and create a new set of merge decisions in the deeper level (with its "virtual base" being those changes that appear in both elements). 

Example structures of two inserted cells:

Local cell:
```json
{
    "source": [
        "# Copyright Bob, 2016\n",
        "import os\n",
        "\n",
        "print(os.environ)"
    ],
    "outputs": []
}
```

Remote cell:
```json
{
    "source": [
        "import os\n",
        "\n",
        "print(os.environ)"
    ],
    "outputs": [
        {
            "output_type" : "stream",
            "name" : "stdout",
            "text" : "[output of os.environ here]",
        }
    ]
}
```

The decisions list would then look like this (if cells were inserted at index 1):
```json
[
    {
        "common_path": "/cells",
        "local_diff": {
            "op": "addrange",
            "key": 1,
            "valuelist": [
                "<local cell as above>"
            ]
        },
        "remote_diff": {
            "op": "addrange",
            "key": 1,
            "valuelist": [
                "<remote cell as above>"
            ]
        },

        "conflict": true,
        "action": "recurse",
        "recursion_base_diff": [
            {
                "op": "addrange",
                "key": 1,
                "valuelist": [
                    {
                        "source": [
                            "import os\n",
                            "\n",
                            "print(os.environ)"
                        ],
                        "outputs": []
                    }
                ]
            }
        ],
        "recursion_decisions": [
            {
                "common_path": "/cells/1/source",
                "local_diff": [
                    {
                        "op": "addrange",
                        "key": 0,
                        "valuelist": "# Copyright Bob, 2016\n"
                    }
                ],
                "action": "local"
            },
            {
                "common_path": "/cells/1/outputs",
                "remote_diff": [
                    {
                        "op": "addrange",
                        "key": 0,
                        "valuelist": [
                            {
                                "output_type" : "stream",
                                "name" : "stdout",
                                "text" : "[output of os.environ here]",
                            }
                        ]
                    }
                ],
                "action": "remote"
            }
        ]
    }
]
```

Thoughts?

**Comment by vidartf** [comment]
One unexpected issue with using `common_path`: Any keys with slashes in them screws up its resolution, e.g. "plain/text". I hesitate to add an escape/unescape mechanism, because they be a hassle to deal with. We could of course try to use another path separator in `common_path`, but as far as I can tell there are no restrictions on key names in nbformat, so we will always have this issue (even escape sequences are not guaranteed to be unique...).

**Comment by martinal** [comment]
Could always use a tuple in python at least, it's immutable and can be used as dict key.

**Comment by vidartf** [comment]
The JSON equivalent of that would be to change `common_path` to be a list of strings instead of a single string. That can then of course be translated to a tuple in the python code. I'll give a go and see if that is easy to deal with before updating this PR.

**Comment by vidartf** [comment]
It seems to work nicely, so I've updated the schema.

**Comment by vidartf** [comment]
Implemented in #76.
